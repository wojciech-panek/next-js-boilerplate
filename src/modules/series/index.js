export {
  SeriesTypes,
  SeriesActions,
} from './series.redux';

export * from './series.selectors';
