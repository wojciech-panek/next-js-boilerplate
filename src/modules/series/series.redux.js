import { createActions, createReducer } from 'reduxsauce';
import { Record, List, Map, fromJS } from 'immutable';


export const { Types: SeriesTypes, Creators: SeriesActions } = createActions({
  fetch: [],
  fetchSuccess: ['data'],
  fetchError: ['payload'],
  fetchSingle: ['id'],
  fetchSingleSuccess: ['data'],
  fetchSingleError: ['payload'],
}, { prefix: 'SERIES_' });

const SeriesRecord = new Record({
  items: List(),
  data: Map(),
});

export const INITIAL_STATE = new SeriesRecord({});

const getSuccessHandler = (state = INITIAL_STATE, action) => state.set('items', fromJS(action.data));

const getSingleSuccessHandler = (state = INITIAL_STATE, action) => state.set('data', fromJS(action.data));

export const reducer = createReducer(INITIAL_STATE, {
  [SeriesTypes.FETCH_SUCCESS]: getSuccessHandler,
  [SeriesTypes.FETCH_SINGLE_SUCCESS]: getSingleSuccessHandler,
});
