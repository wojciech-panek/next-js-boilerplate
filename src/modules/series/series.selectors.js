import { createSelector } from 'reselect';


const selectSeriesDomain = state => state.get('series');

export const selectSeriesItems = createSelector(
  selectSeriesDomain, state => state.get('items')
);

export const selectSeriesData = createSelector(
  selectSeriesDomain, state => state.get('data')
);
