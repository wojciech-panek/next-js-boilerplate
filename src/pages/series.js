import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'next/router';
import { compose } from 'ramda';

import withIntl from '../shared/lib/withIntl';


export class SeriesPage extends PureComponent {
  static propTypes = {
    router: PropTypes.object.isRequired,
  };

  render() {
    return (
      <div>Series: {this.props.router.query.id}</div>
    );
  }
}

export default compose(
  withIntl,
  withRouter,
)(SeriesPage);
