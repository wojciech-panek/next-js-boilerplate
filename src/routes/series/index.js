export { default as Series } from './series.container';
