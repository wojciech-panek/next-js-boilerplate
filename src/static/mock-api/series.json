[
  {
    "id": 1,
    "title": "Dick in Ibiza",
    "season": "1"
  },
  {
    "id": 2,
    "title": "Dick in Ibiza",
    "season": "2"
  },
  {
    "id": 3,
    "title": "Liars",
    "season": "1"
  },
  {
    "id": 4,
    "title": "F*ck My Life",
    "season": "1"
  }
]
