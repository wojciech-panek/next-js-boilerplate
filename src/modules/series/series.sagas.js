import { put, takeLatest } from 'redux-saga/effects';
import { delay } from 'redux-saga';

import reportError from '../../shared/utils/reportError';
import api from '../../shared/services/api';
import { SeriesTypes, SeriesActions } from './series.redux';


export function* fetchSeries() {
  try {
    yield delay(1000);
    const { data } = yield api.get('/static/mock-api/series.json');

    return yield put(SeriesActions.fetchSuccess(data));
  } catch (e) {
    if (e.response) {
      return yield put(SeriesActions.fetchError(e.response.data));
    }

    return yield reportError(e);
  }
}

export function* fetchSingleSeries({ id }) {
  try {
    yield delay(1000);
    const { data: seriesData } = yield api.get('/static/mock-api/series.json');
    const { data: episodesData } = yield api.get('/static/mock-api/episodes.json');

    const data = seriesData.find((series) => series.id === id);
    data.episodes = episodesData.filter((episide) => episide.series === id);

    return yield put(SeriesActions.fetchSingleSuccess(data));
  } catch (e) {
    if (e.response) {
      return yield put(SeriesActions.fetchSingleError(e.response.data));
    }

    return yield reportError(e);
  }
}

export function* watchSeries() {
  try {
    yield takeLatest(SeriesTypes.FETCH, fetchSeries);
    yield takeLatest(SeriesTypes.FETCH_SINGLE, fetchSingleSeries);
  } catch (e) {
    yield reportError(e);
  }
}
