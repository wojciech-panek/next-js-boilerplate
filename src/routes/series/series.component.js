import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Head from 'next/head';

import { Loader } from '../../shared/components/loader';
import { Link } from '../../shared/components/link';
import { H2, H4 } from '../../shared/theme/typography';
import { Container, EpisodesContainer } from './series.styles';
import messages from './series.messages';



export class Series extends PureComponent {
  static propTypes = {
    data: PropTypes.object,
    router: PropTypes.object.isRequired,
  };

  getData = () => {
    if (this.props.data.size > 0 && this.props.data.get('id') === parseInt(this.props.router.query.id, 10)) {
      return this.props.data;
    }
    return null;
  };

  renderEpisode = (episode) => {
    return (
      <Link key={episode.get('id')} to={`/episodes/${episode.get('id')}`}>
        <a>[{episode.get('number')}] {episode.get('title')}</a>
      </Link>
    );
  };


  render() {
    const data = this.getData();

    if (!data) {
      return <Loader />;
    }

    return (
      <Container>
        <FormattedMessage {...messages.pageTitle}>
          {pageTitle => (
            <Head>
              <title key="title">{ pageTitle }</title>
            </Head>
          )}
        </FormattedMessage>

        <H2>{data.get('title')} [<FormattedMessage {...messages.season} />: {data.get('season')}]</H2>

        <H4><FormattedMessage {...messages.episodes} />:</H4>

        <EpisodesContainer>
          {data.get('episodes').toArray().map(this.renderEpisode)}
        </EpisodesContainer>

        <Link to="/">
          <a><FormattedMessage {...messages.back} /></a>
        </Link>
      </Container>
    );
  }
}
