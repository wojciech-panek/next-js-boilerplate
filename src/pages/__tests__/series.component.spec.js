import React from 'react';
import { shallow } from 'enzyme';

import { SeriesPage } from '../series';


describe('Series: Page', () => {
  const defaultProps = {
    router: { query: { id: 'test-id' } },
  };

  const component = (props) => (
    <SeriesPage {...defaultProps} {...props} />
  );

  it('should render correctly', () => {
    const wrapper = shallow(component());
    global.expect(wrapper).toMatchSnapshot();
  });
});
