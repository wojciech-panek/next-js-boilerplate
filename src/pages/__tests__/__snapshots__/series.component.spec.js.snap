// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Series: Page should render correctly 1`] = `
<Series
  router={
    Object {
      "query": Object {
        "id": "test-id",
      },
    }
  }
/>
`;
